wisp-sampling/src/main/kotlin/wisp/sampling/Sampler.kt

```diff
@@ -0,0 +1,72 @@
+package wisp.sampling
+
+import java.util.concurrent.ThreadLocalRandom
+import java.util.concurrent.TimeUnit
+
+/**
+ * A [Sampler] is used to "sample" whether an action is allowed to occur or not.
+ *
+ * A common usage of the sampler would look like:
+ *
+ * ```kt
+ * if (sampler.sample()) {
+ *     performAction()
+ * }
+ * ```
+ *
+ * The frequency at which `sample` returns `true` or `false` is based on the implementation's policy. For example,
+ * [Sampler.always] creates a [Sampler] that only returns `true`, while [Sampler.percentage] will only return `true` for a
+ * given percentage of samples. For a more complex example, [Sampler.rateLimiting] will limit the number of `true`
+ * samples to a given rate per second.
+ */
+interface Sampler {
+    /**
+     * Tests if a sample is allowed or not
+     *
+     * @return `true` if sample is allowed, otherwise `false`
+     */
+    fun sample(): Boolean
+
+    companion object {
+        /**
+         * Creates a [Sampler] that limits positive results to a percentage chance.
+         *
+         * @param samplePercentage percentage chance of positive results
+         *
+         * @return percentage [Sampler] instance
+         */
+        fun percentage(samplePercentage: Int): Sampler = PercentageSampler(samplePercentage) {
+            ThreadLocalRandom.current().nextInt(0, 100)
+        }
+
+        /**
+         * Creates a [Sampler] that limits positive results to a limited rate per second.
+         *
+         * @param ratePerSecond the number of positive results per second
+         *
+         * @return rate limiting [Sampler] instance
+         */
+        fun rateLimiting(ratePerSecond: Long): Sampler = RateLimitingSampler(RateLimiter(ratePerSecond))
+
+        /**
+         * Creates a [Sampler] that always returns positive results.
+         *
+         * @return always [Sampler] instance
+         */
+        fun always(): Sampler = AlwaysSampler()
+    }
```

Comment:
Another change from the [misk version](https://github.com/cashapp/misk/blob/master/misk-core/src/main/kotlin/misk/sampling/Sampler.kt), mostly so that I could mark the implementations as `internal` to reduce the API footprint to just the `Sampler` interface.